### PR TITLE
Refactor InitSteps and INethermindApi to accomodate factories

### PR DESCRIPTION
**Pull request by jmederosalvarado**
There is a "problem" in the way we do initialization right now. Namely, we hardcode the type of many important components like `BlockProcessor` in the initialization steps, which makes it very hard for plugins to override such components.

The workaround is what we do in AuRa. We create custom initialization steps and override the parts that we need to customize. Overriding init steps is not ideal, since Init steps are a complex part of the client, and doing it correctly is not simple.

The approach proposed in this PR is to have a set of factories in the NethermindApi. These factories would be initialized with the default types in a step before initializing plugins. Plugins would then have a chance to override these factories so that the next steps are able to use the customize types.

## Changes

- Create new InitializeFactories InitSteps
- Create new IApiWithFactories
- Refactor initialization logic in other steps to use factories

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

**Comment by jmederosalvarado** [comment]
The approach described was only followed with BlockProcessor. If we decide that it makes sense I could further extend it to other components (especially the ones initialized on the InitializeBlockchain step).

File: src/Nethermind/Nethermind.Api/NethermindApi.cs [hunk 1]
```diff
@@ -100,6 +101,8 @@ public IBlockchainBridge CreateBlockchainBridge()
             );
         }
 
+        public IBlockProcessorFactory BlockProcessorFactory { get; set; }
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.Api/NethermindApi.cs:104]: I would like to separate the factories from the rest of the APIs.
something like this
```
NethermindApi.Factories.BlockProcessingFactory
```

**Review by smartprogrammer93 — APPROVED** [review]
I really like this

**Review by asdacap — APPROVED** [review]

**Comment by smartprogrammer93** [comment]
I believe this resolves #3486, since the plugin can inject logic before and after the initialization through the factory

**Comment by jmederosalvarado** [comment]
Thanks @smartprogrammer93 . I was using this PR recently to implement a plugin and I noticed that there is still kind of a deficiency. The arguments we pass to the factory's create method are still hardcoded. I'm thinking we might want to do something like dependency injection, where we use reflection to check the factory's create arguments and inject them from the NethermindApi. WDYT?

**Comment by asdacap** [comment]
Careful with over complicating things. Modern DI framework can do these and much more. The problem is the configuration can get much more complicated than necessary, too much `magic` and a lot of the time, too many people in the team/outside the team/new to the team can't handle it. I recommend just modifing things when needed.

**Comment by smartprogrammer93** [comment]
With DI, we wont need the factories in the first place because the DI engine will be resolving the registered services. A plugin can simply replace a registered service in the DI before the initialization of those services. Unfortunately, I have suggested to the leads that we switch to a DI engine instead of this NethermindApi approach, the answer was it is easier said than done which i agree with now since i have seen more of the code. I believe the factories in their hard coded form could be a step in the right direction and then bit by bit we can start switching to fully using a DI.

**Comment by smartprogrammer93** [comment]
@asdacap I dont think DI are `magic` you register a service, it gets populated when needed. it is simple. It is a pattern used in most modern applications especially in the .Net ecosystem.

**Comment by LukaszRozmej** [comment]
Truth be told I don't have a strong preference. Question from me does this actually solve an ACTUAL problem, not only a hipothetical one?

Dependency Injection might be the end-goal, but with as complex an initialization as ours, would it work well? How many regressions would we introduce? Hard to say without experimenting with it - and not sure if potential gain is big enough for now.

**Comment by jmederosalvarado** [comment]
@asdacap @smartprogrammer93 I agree that a full-blown DI engine maybe is overcomplicating things (maybe?). I would suggest a simpler in-house approach, we register types somewhere in the Api. When we need to initialize them we check constructor arguments using reflection and try to match them using what we have already initialized in the NethermindApi.

An alternative approach would be to pass the whole NethermindApi to the factory's create method, allowing it to use whatever it needs. Maybe this is better.

**Comment by jmederosalvarado** [comment]
> does this actually solve an ACTUAL problem, not only a hipothetical one?

@LukaszRozmej it does solve an actual problem, one of the main reasons why AuRa-related plugins have custom initialization steps is because you have no step where you can override the core components in the Api in a simpler, more plugin-friendly manner. I just came across the same problem while trying to prototype Polygon support in Nethermind. In general anything that requires major customization will need to do this and right now we only have hacky workarounds or complex solutions.

**Comment by smartprogrammer93** [comment]
> @asdacap @smartprogrammer93 I agree that a full-blown DI engine maybe is overcomplicating things (maybe?). I would suggest a simpler in-house approach, we register types somewhere in the Api. When we need to initialize them we check constructor arguments using reflection and try to match them using what we have already initialized in the NethermindApi.
> 
> An alternative approach would be to pass the whole NethermindApi to the factory's create method, allowing it to use whatever it needs. Maybe this is better.

Would have to dissgree here. Why reinvent the wheel? This is something that has been used and developed for so long, with already massive use cases. What about managing the lifecycle of services, i understand that in our client, 99% of services is a singleton, but is that really the correct approach? For certain use cases, possibly yes, but for all, definitely not.

**Comment by smartprogrammer93** [comment]
Yes @jmederosalvarado, there is this one from microsoft that i like the most. 
https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection?view=dotnet-plat-ext-7.0
We can still use an iterative approach with it without introducing regressions imo.

**Pull request closed without merging**